eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -140,264 +433,245 @@ proposalsAddVote voter vote gai (Proposals omap) =
       StakePoolVoter kh -> insertVote gasStakePoolVotesL kh
       CommitteeVoter c -> insertVote gasCommitteeVotesL c
 
--- | Extract `GovActionState`s for the given set of `GovActionId`s from the `Proposals`
+-- | For each action-id in the given set. attempt to remove it from
+-- the @`Proposals`@ forest based on its purpose. Although the removal
+-- operations are applied to parts of the forest without any checks, we
+-- cover them in property-tests
 proposalsRemoveIds ::
+  forall era.
   Set (GovActionId (EraCrypto era)) ->
   Proposals era ->
   (Proposals era, Map.Map (GovActionId (EraCrypto era)) (GovActionState era))
-proposalsRemoveIds gais (Proposals omap) =
-  let (retained, removed) = OMap.extractKeys gais omap
-   in (Proposals retained, removed)
-
--- | Extract `GovActionState`s for all the descendents of a set of
--- `GovActionId`s from the `Proposals`
+proposalsRemoveIds gais ps =
+  let (retainedOMap, removedFromOMap) = OMap.extractKeys gais $ ps ^. pPropsL
+      (roots, hierarchy) = foldl' removeEach (ps ^. pRootsL, ps ^. pGraphL) gais
```

Comment:
This should let us avoid the lookup couple of lines below, because we'll be iterating over the actual states, not the ids.
```suggestion
      (roots, hierarchy) = foldl' removeEach (ps ^. pRootsL, ps ^. pGraphL) removedFromOMap
```